rp2040-hal/src/pio.rs

```diff
@@ -656,22 +726,89 @@ pub struct Tx<SM: ValidStateMachine> {
 }
 
 impl<SM: ValidStateMachine> Tx<SM> {
+    fn block(&self) -> &pac::pio0::RegisterBlock {
+        // Safety: The register is unique to this Tx instance.
+        unsafe { &*self.block }
+    }
+
     /// Write an element to TX FIFO.
     ///
     /// Returns `true` if the value was written to FIFO, `false` otherwise.
-    pub fn write(&mut self, value: u32) -> bool {
+    pub fn write<T>(&mut self, value: T) -> bool {
         // Safety: The register is never written by software.
-        let is_full = unsafe { &*self.block }.fstat.read().txfull().bits() & (1 << SM::id()) != 0;
+        let is_full = self.is_full();
 
         if is_full {
             return false;
         }
 
-        // Safety: The register is unique to this Tx instance.
-        unsafe { &*self.block }.txf[SM::id()].write(|w| unsafe { w.bits(value) });
+        unsafe {
+            let reg_ptr = self.block().txf[SM::id()].as_ptr() as *mut T;
+            core::ptr::write_volatile(reg_ptr, value);
+        }
 
         true
     }
+
+    /// Checks if the state machine has stalled on empty TX FIFO during a blocking PULL, or an OUT
+    /// with autopull enabled.
+    pub fn has_stalled(&self) -> bool {
```

Comment:
> I'd prefer devsnek's suggestion over just "stalled". it's also more accurate here - has_stalled could mean that it may or may not currently be stalled, but has stalled some time since you last asked/cleared a flag.
Isn't that an argument in favor of `has_stalled`?
The flag is documented as "State machine has stalled on empty TX FIFO during ablocking PULL, or an OUT with autopull enabled. Write 1 to clear."
So its meaning is, as you wrote, "that it may or may not currently be stalled, but has stalled some time since you last asked/cleared a flag".